### PR TITLE
Automated Functional Test of BS: request download exchange set

### PR DESCRIPTION
**Pull request by Arit14975**
Hi All,

This PR consist of FT  and required mock changes under [Product Backlog Item 150897](https://dev.azure.com/ukhydro/Exchange%20Set%20Service/_workitems/edit/150897): Testing : BESS BS - Request, wait/poll and download exchange set
- The changes made to the FT Job tasks, such as adding variables and adjusting the sequence in yaml file to run FT's in pipeline.

### CODE REVIEW

**Review by Mayuresh-Gawde — APPROVED** [review]
LGTM from QA!!

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/TestData/ConfigFile.json [hunk 1]
```diff
@@ -3,10 +3,11 @@
...
     "isEncrypted": false,
     "encCellNames": [
-        "GB123456",
-        "GB234567",
-        "GB*",
-        "GB1*"
+        //"GB123456",
+        //"GB234567",
+        //"GB*",
+        //"GB1*"
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/TestData/ConfigFile.json:9]: Can we remove commented lines.

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/TestConfiguration.cs [hunk 2]
```diff
@@ -17,36 +21,61 @@ public class BessApiConfiguration
...
+        {
+            public string? BaseUrl { get; set; }
+            public int BatchCommitWaitTime { get; set; }
+        }
+
+        public class BessStorageConfiguration
+        {
+            public string? QueueName { get; set; }
+        }
+        public class ExchangeSetDetails
```
Comment (harshal11869) [inline UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/TestConfiguration.cs:56]:
```suggestion
        public class BessStorageConfiguration
        {
            public string? QueueName { get; set; }
        }
        
        public class ExchangeSetDetails
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/TestData/ConfigFile.json (hunk 1 above)
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/TestData/ConfigFile.json]: Resolved.  Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/TestConfiguration.cs (hunk 2 above)
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/TestConfiguration.cs]: Resolved.  Thanks

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by amogh14893 — APPROVED** [review]
LGTM

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/EssEndpointHelper.cs [hunk 3]
```diff
@@ -46,12 +44,28 @@ public static async Task<HttpResponseMessage> ProductVersionsEndpoint(string? ba
...
+
+        /// <summary>
+        /// This Method is used to check the Batch Status and Download ES
+        /// </summary>
+        /// <param name="batchId">Provide the BatchId of the requested Product</param>
+        /// <returns></returns>
+        public static async Task<string> CreateExchangeSetFile(string batchId)
+        {
+            // The below sleep is to give time to BuilderService to download the required zip
+            Thread.Sleep(90000);
```
Comment (DanOwenUKHO) [inline UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/EssEndpointHelper.cs:60]: Is this in the required spot? It doesn't make sense to sleep the thread right at the start of the method?

**Review by DanOwenUKHO — COMMENTED** [review]
Some comments/questions

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/TestData/ConfigFile.json [hunk 4]
```diff
@@ -1,12 +1,8 @@
 {
     "name": "BES-1",
     "exchangeSetStandard": "s63",
-    "isEncrypted": false,
     "encCellNames": [
-        "GB123456",
-        "GB234567",
-        "GB*",
-        "GB1*"
+        "GB301910"
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/TestData/ConfigFile.json:5]: Hi Dan, For now, we just need any cell to test. So we have removed the rest of them. As and when required, we will add those. Thanks.

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            while (DateTime.UtcNow - startTime < TimeSpan.FromMinutes(config.BatchCommitWaitTime))
+            {
+                HttpResponseMessage batchStatusResponse = await FssApiClient.GetBatchStatusAsync(batchStatusUri);
+                batchStatusResponse.StatusCode.Should().Be((HttpStatusCode)200);
+                
+                var batchStatusResponseObj = JsonConvert.DeserializeObject<ResponseBatchStatusModel>(await batchStatusResponse.Content.ReadAsStringAsync());
+                batchStatus = batchStatusResponseObj!.Status!;
+
+                if (batchStatus!.Equals("Committed"))
+                    break;
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs:41]: Resolved. Thanks

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/EssEndpointHelper.cs (hunk 3 above)
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/EssEndpointHelper.cs]: Thanks Dan, We have moved that to a new method.

File: UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            var startTime = DateTime.UtcNow;
+            while (DateTime.UtcNow - startTime < TimeSpan.FromMinutes(config.BatchCommitWaitTime))
+            {
+                HttpResponseMessage batchStatusResponse = await FssApiClient.GetBatchStatusAsync(batchStatusUri);
+                batchStatusResponse.StatusCode.Should().Be((HttpStatusCode)200);
+                
+                var batchStatusResponseObj = JsonConvert.DeserializeObject<ResponseBatchStatusModel>(await batchStatusResponse.Content.ReadAsStringAsync());
+                batchStatus = batchStatusResponseObj!.Status!;
+
+                if (batchStatus!.Equals("Committed"))
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.BESS.API.FunctionalTests/Helpers/FssBatchHelper.cs]: Resolved. Thanks

**Review by DanOwenUKHO — APPROVED** [review]

**Pull request merged**